Add form for `AuraRuleElement`

In3luki opened the pull request:
![Basic](https://github.com/foundryvtt/pf2e/assets/41452412/8fe4f1de-6eea-440e-9351-b8be3eae83c1)

![Effects](https://github.com/foundryvtt/pf2e/assets/41452412/64d6ea50-5ce7-4a6f-a47c-8a9ed3515d66)

![Appearance1](https://github.com/foundryvtt/pf2e/assets/41452412/6177c0dc-1561-49bf-9480-02823f8bfebd)

![Appearance2](https://github.com/foundryvtt/pf2e/assets/41452412/7bf2bbc6-e8bf-4245-86b5-94b550e997a6)

stwlam commented:
The positioning of the checkboxes and their respective labels is a little confusing

stwlam commented on static/templates/items/rules/aura.hbs:
```diff
@@ -0,0 +1,231 @@
+<div data-rule-tab="basic">
+    <div class="form-group">
+        <div class="spaced-fields details-container-flex-row">
+            <label>{{localize "PF2E.RuleEditor.Aura.Basic.Radius"}}</label>
```
Set these labels in the schema (see MultipleAttackPenalty)

In3luki commented:
![image](https://github.com/foundryvtt/pf2e/assets/41452412/e28609ab-b809-46a6-8be4-76eae60581dc)

stwlam commented:
Can we exclude (or at least hide for now) any part of the RE that doesn't do anything yet? Saves and events are the two I'm seeing.

In3luki commented:
I've started to remove events and saves and now the whole grid layout falls apart. :upside_down_face: 
I'll redo the whole grid with a stricter layout for labels and content tomorrow.
I think moving from two options per row to one might be better now.

In3luki commented:
![image](https://github.com/foundryvtt/pf2e/assets/41452412/edf9394b-01e6-4566-8e96-dbbd154344cd)

![image](https://github.com/foundryvtt/pf2e/assets/41452412/628c8a09-0021-49bb-b852-d7faaa8fcdbd)

![image](https://github.com/foundryvtt/pf2e/assets/41452412/b875d611-01c4-442a-a76e-59f5c1b01d19)

stwlam commented:
Conflicts!

In3luki commented:
Updated to solve merge conflicts. I've also moved `cleanDataUsingSchema` and `cleanPredicate` back to being functions because they are not needed in the subclass after all.

CarlosFdez commented on src/module/item/sheet/rule-element-form/aura.ts:
```diff
@@ -0,0 +1,202 @@
+            const fieldset = htmlClosest(button, "fieldset");
+            const select = htmlQuery<HTMLSelectElement>(fieldset, "select");
+            const input = htmlQuery<HTMLInputElement>(fieldset, "input");
+            if (select && input) {
+                if (!select.value) {
+                    button.disabled = true;
+                    input.disabled = true;
+                }
+            }
```
```suggestion
            if (select && input && !select.value) {
                button.disabled = true;
                input.disabled = true;
            }
```

stwlam commented on static/templates/items/rules/aura.hbs:
```diff
@@ -0,0 +1,208 @@
+                    >
+                </div>
+
+                <div class="two-items long-label">
+                    <label>
+                        {{localize fields.appearance.fields.texture.fields.playbackRate.label}}
+                        <i class="fas fa-info-circle small" data-tooltip={{fields.appearance.fields.texture.fields.playbackRate.hint}}></i>
+                    </label>
+                    {{rangePicker name=(concat "system.rules." index ".appearance.texture.playbackRate") value=object.appearance.texture.playbackRate placeholder="1" min="0" max="4" step="0.05"}}
+                </div>
+
+                <div class="grid-item long-label">
+                    <label>
+                        {{localize fields.appearance.fields.texture.fields.loop.label}}
+                        <i class="fas fa-info-circle small" data-tooltip={{fields.appearance.fields.texture.fields.loop.hint}}></i>
+                    </label>
+                    <input type="checkbox" name="system.rules.{{index}}.appearance.texture.loop" {{checked object.appearance.texture.loop}}/>
+                </div>
```
Can these be wise to the file path and be disabled when it's an image?

stwlam commented on static/templates/items/rules/aura.hbs:
```diff
@@ -0,0 +1,210 @@
+                    >
+                </div>
+
+                {{#if isVideoFile}}
```
This should be a check that it's not an image file path since it's injectable. Do also remove the values or at least reset them to defaults when they're not applicable.

stwlam commented on src/module/item/sheet/rule-element-form/aura.ts:
```diff
@@ -0,0 +1,210 @@
+            this.object = new AuraRuleElement(this.rule, { parent: item });
+        }
+
+        this.#effectsMap.clear();
+        this.object.toObject().effects.forEach((effect, index) => {
+            this.#effectsMap.set(`${index}`, effect);
+        });
```
No `forEach` :blush:

stwlam replied:
```suggestion
        this.#effectsMap = new Map(this.object.toObject().effects.map((e, index): [number, AuraEffectSource] => [index, e]));
```

stwlam commented on src/module/item/sheet/rule-element-form/aura.ts:
```diff
@@ -0,0 +1,210 @@
+        displayStyle: "grid",
+    };
+
+    #effectsMap: Map<string, AuraEffectSource> = new Map();
```
```suggestion
    #effectsMap: Map<number, AuraEffectSource> = new Map();
```

stwlam commented on src/module/item/sheet/rule-element-form/aura.ts:
```diff
@@ -0,0 +1,210 @@
+
+    #addEffect(uuid: ItemUUID): void {
+        const index = this.#effectsMap.size + 1;
+        this.#effectsMap.set(`${index}`, { uuid } as AuraEffectSource);
```
```suggestion
        this.#effectsMap.set(index, { uuid } as AuraEffectSource);
```

stwlam commented on static/lang/re-en.json:
```diff
@@ -1028,6 +1028,61 @@
         "RuleEditor": {
+            "Aura": {
+                "Appearance": {
+                    "Alpha": "Opacity",
```
```suggestion
```

stwlam commented on src/module/rules/rule-element/aura.ts:
```diff
@@ -100,65 +117,154 @@ class AuraRuleElement extends RuleElementPF2e<AuraSchema> {
+                        required: true,
+                        nullable: false,
+                        initial: 0.75,
+                        label: "PF2E.RuleEditor.Aura.Appearance.Alpha",
```
```suggestion
                        label: "PF2E.RuleEditor.General.Opacity",
```

stwlam commented on src/module/rules/rule-element/aura.ts:
```diff
@@ -100,65 +117,154 @@ class AuraRuleElement extends RuleElementPF2e<AuraSchema> {
+                        required: false,
+                        nullable: false,
+                        initial: 0.25,
+                        label: "PF2E.RuleEditor.Aura.Appearance.Alpha",
```
```suggestion
                        label: "PF2E.RuleEditor.General.Opacity",
```

stwlam commented on src/module/rules/rule-element/aura.ts:
```diff
@@ -100,65 +117,154 @@ class AuraRuleElement extends RuleElementPF2e<AuraSchema> {
+                        nullable: false,
+                        positive: true,
+                        initial: 1,
+                        label: "PF2E.RuleEditor.Aura.Appearance.Scale",
```
```suggestion
                        label: "Scale",
```

stwlam commented on src/module/rules/rule-element/aura.ts:
```diff
@@ -100,65 +117,154 @@ class AuraRuleElement extends RuleElementPF2e<AuraSchema> {
                         required: false,
                         nullable: false,
                         initial: undefined,
+                        label: "PF2E.RuleEditor.Aura.Appearance.Translation.Label",
+                        hint: "PF2E.RuleEditor.Aura.Appearance.Hints.Translation",
```
Put these in the same object

stwlam commented on src/module/rules/rule-element/aura.ts:
```diff
@@ -100,65 +117,154 @@ class AuraRuleElement extends RuleElementPF2e<AuraSchema> {
                         positive: true,
                         max: 4,
                         initial: 1,
+                        label: "PF2E.RuleEditor.Aura.Appearance.PlaybackRate",
+                        hint: "PF2E.RuleEditor.Aura.Appearance.Hints.PlaybackRate",
```
Same

stwlam commented on src/module/rules/rule-element/aura.ts:
```diff
@@ -100,65 +117,154 @@ class AuraRuleElement extends RuleElementPF2e<AuraSchema> {
+                        required: false,
+                        nullable: false,
+                        initial: undefined,
+                        label: "PF2E.RuleEditor.Aura.Appearance.Loop",
+                        hint: "PF2E.RuleEditor.Aura.Appearance.Hints.Loop",
```
Same

stwlam commented on static/lang/re-en.json:
```diff
@@ -1028,6 +1028,61 @@
+                    "Label": "Appearance",
+                    "Loop": "Loop",
+                    "PlaybackRate": "Playback Rate",
+                    "Scale": "Scale",
+                    "Src": "File",
```
```suggestion
```

stwlam commented on src/module/rules/rule-element/aura.ts:
```diff
@@ -100,65 +117,154 @@ class AuraRuleElement extends RuleElementPF2e<AuraSchema> {
+                        required: true,
+                        nullable: false,
+                        initial: undefined,
+                        label: "PF2E.RuleEditor.Aura.Appearance.Src",
```
```suggestion
                        label: "TOKEN.ImagePath",
```

In3luki commented:
Image path is a full row now.  All other values still show up correctly.

![image](https://github.com/foundryvtt/pf2e/assets/41452412/97fee64c-2112-49f2-8993-dda25f908ed4)
